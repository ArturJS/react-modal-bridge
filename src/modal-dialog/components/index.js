export * from './custom-type.jsx';
export * from './standard-type.jsx';
