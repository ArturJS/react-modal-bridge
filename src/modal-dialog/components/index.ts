export * from './base-modal';
export * from './custom-type'; // eslint-disable-line import/no-cycle

export * from './standard-type'; // eslint-disable-line import/no-cycle