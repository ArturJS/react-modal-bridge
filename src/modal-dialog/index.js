export * from './modal-dialog.jsx'; // eslint-disable-line import/no-cycle
