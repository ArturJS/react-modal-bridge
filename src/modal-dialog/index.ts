export * from './modal-dialog'; // eslint-disable-line import/no-cycle