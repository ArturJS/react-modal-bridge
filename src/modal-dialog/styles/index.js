import './buttons.scss';
import './main.scss';
