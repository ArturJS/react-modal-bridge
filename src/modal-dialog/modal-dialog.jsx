// @flow
import React, { memo, useState, useEffect } from 'react';
import ReactModal from 'react-modal';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import {
  MODAL_TYPES,
  CLOSE_DELAY_MS,
  Modal,
  modalService
} from '../modal.service';
import { CustomType, StandardType } from './components';
import './styles';

const defaultBackdropStyle = {
  overlay: {
    backgroundColor: ''
  }
};

const noBackdropStyle = {
  overlay: {
    backgroundColor: 'transparent',
    pointerEvents: 'none',
    zIndex: 1080
  }
};

const useModalsSubscription = () => {
  const [modals, setModals] = useState([]);

  useEffect(() => {
    const unsubscribe = modalService.subscribe(
      // eslint-disable-next-line no-shadow
      ({ modals }: {| modals: Array<Modal> |}) => {
        setModals(modals);
      }
    );

    return unsubscribe;
  }, []);

  return modals;
};

export const ModalDialog = memo(() => {
  const dismiss = (id: number) => {
    modalService.dismiss({
      id
    });
  };

  const modals = useModalsSubscription();

  return (
    <>
      {modals.map(modal => (
        <ReactModal
          key={modal.id}
          isOpen={modal.isOpen}
          onRequestClose={() => dismiss(modal.id)}
          style={modal.noBackdrop ? noBackdropStyle : defaultBackdropStyle}
          className={`rmb-modal ${modal.className}`}
          shouldCloseOnOverlayClick={modal.shouldCloseOnOverlayClick}
          closeTimeoutMS={CLOSE_DELAY_MS}
          contentLabel=""
          ariaHideApp={false}
        >
          <TransitionGroup>
            {modal.isOpen && (
              <CSSTransition
                key={modal.id}
                appear
                timeout={CLOSE_DELAY_MS}
                classNames="rmb-modal-show"
                mountOnEnter
                unmountOnExit
              >
                <div className="rmb-modal-content">
                  <button
                    type="button"
                    className="rmb-close"
                    onClick={() => dismiss(modal.id)}
                  >
                    &times;
                  </button>
                  <div className="rmb-modal-header">
                    <h3 className="rmb-modal-title">{modal.title}</h3>
                  </div>
                  {modal.type === MODAL_TYPES.custom && (
                    <CustomType modal={modal} />
                  )}
                  {modal.type !== MODAL_TYPES.custom && (
                    <StandardType modal={modal} />
                  )}
                </div>
              </CSSTransition>
            )}
          </TransitionGroup>
        </ReactModal>
      ))}
    </>
  );
});
