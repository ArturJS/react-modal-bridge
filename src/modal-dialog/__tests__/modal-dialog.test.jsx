import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { modalService } from '../../modal.service';
import { ModalDialog } from '../modal-dialog.jsx';

// eslint-disable-next-line react/prop-types
jest.mock('react-modal', () => ({ children, ...props }) => (
  <div className="react-modal-mock">
    <pre className="props">{JSON.stringify(props, null, 2)}</pre>
    {children}
  </div>
));

jest.mock('react-transition-group', () => ({
  // eslint-disable-next-line react/prop-types
  CSSTransition: ({ children, ...props }) => (
    <div className="CSSTransition-mock">
      <pre className="props">{JSON.stringify(props, null, 2)}</pre>
      {children}
    </div>
  ),
  // eslint-disable-next-line react/prop-types
  TransitionGroup: ({ children, ...props }) => (
    <div className="TransitionGroup-mock">
      <pre className="props">{JSON.stringify(props, null, 2)}</pre>
      {children}
    </div>
  )
}));

const flushPromises = () => new Promise(resolve => setImmediate(resolve));

describe('<ModalDialog />', () => {
  describe('type `confirm`', () => {
    afterEach(async () => {
      await modalService.closeAll();
    });

    it('should match snapshot', () => {
      modalService.confirm({ title: 'Confirm title', body: 'Confirm body' });

      const tree = renderer.create(<ModalDialog />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should contain modal content', () => {
      modalService.confirm({ title: 'Confirm title', body: 'Confirm body' });

      const wrapper = mount(<ModalDialog />);

      expect(wrapper.find('.rmb-modal-content').exists()).toBe(true);
    });

    it('should be closed', async () => {
      let isClosed = false;

      modalService
        .confirm({ title: 'Confirm title', body: 'Confirm body' })
        .result.then(() => {
          isClosed = true;
        });

      const wrapper = mount(<ModalDialog />);

      wrapper.find('.rmb-btn-ok').simulate('click');
      await flushPromises();

      expect(wrapper.find('.rmb-modal-content').exists()).toBe(false);
      expect(isClosed).toBe(true);
    });

    it('should be dismissed by clicking [Cancel] button', async () => {
      let isDismissed = false;

      modalService
        .confirm({
          title: 'Confirm title',
          body: 'Confirm body',
          throwCancelError: true
        })
        .result.catch(() => {
          isDismissed = true;
        });

      const wrapper = mount(<ModalDialog />);

      wrapper.find('.rmb-btn-cancel').simulate('click');
      await flushPromises();

      expect(wrapper.find('.rmb-modal-content').exists()).toBe(false);
      expect(isDismissed).toBe(true);
    });

    it('should be dismissed by clicking [X] button', async () => {
      let isDismissed = false;

      modalService
        .confirm({
          title: 'Confirm title',
          body: 'Confirm body',
          throwCancelError: true
        })
        .result.catch(() => {
          isDismissed = true;
        });

      const wrapper = mount(<ModalDialog />);

      wrapper.find('.rmb-close').simulate('click');
      await flushPromises();

      expect(wrapper.find('.rmb-modal-content').exists()).toBe(false);
      expect(isDismissed).toBe(true);
    });
  });

  describe('type `info`', () => {
    afterEach(async () => {
      await modalService.closeAll();
    });

    it('should match snapshot', () => {
      modalService.info({ title: 'Info title', body: 'Info body' });

      const tree = renderer.create(<ModalDialog />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should contain modal content', () => {
      modalService.info({ title: 'Info title', body: 'Info body' });

      const wrapper = mount(<ModalDialog />);

      expect(wrapper.find('.rmb-modal-content').exists()).toBe(true);
    });

    it('should be closed', async () => {
      let isClosed = false;

      modalService
        .info({ title: 'Info title', body: 'Info body' })
        .result.then(() => {
          isClosed = true;
        });

      const wrapper = mount(<ModalDialog />);

      wrapper.find('.rmb-btn-ok').simulate('click');
      await flushPromises();

      expect(wrapper.find('.rmb-modal-content').exists()).toBe(false);
      expect(isClosed).toBe(true);
    });

    it('should be dismissed', async () => {
      let isDismissed = false;

      modalService
        .info({
          title: 'Info title',
          body: 'Info body',
          throwCancelError: true
        })
        .result.catch(() => {
          isDismissed = true;
        });

      const wrapper = mount(<ModalDialog />);

      wrapper.find('.rmb-close').simulate('click');
      await flushPromises();

      expect(wrapper.find('.rmb-modal-content').exists()).toBe(false);
      expect(isDismissed).toBe(true);
    });
  });

  describe('type `error`', () => {
    afterEach(async () => {
      await modalService.closeAll();
    });

    it('should match snapshot', () => {
      modalService.error({ title: 'Error title', body: 'Error body' });

      const tree = renderer.create(<ModalDialog />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should contain modal content', () => {
      modalService.error({ title: 'Error title', body: 'Error body' });

      const wrapper = mount(<ModalDialog />);

      expect(wrapper.find('.rmb-modal-content').exists()).toBe(true);
    });

    it('should be closed', async () => {
      let isClosed = false;

      modalService
        .error({ title: 'Error title', body: 'Error body' })
        .result.then(() => {
          isClosed = true;
        });

      const wrapper = mount(<ModalDialog />);

      wrapper.find('.rmb-btn-ok').simulate('click');
      await flushPromises();

      expect(wrapper.find('.rmb-modal-content').exists()).toBe(false);
      expect(isClosed).toBe(true);
    });

    it('should be dismissed', async () => {
      let isDismissed = false;

      modalService
        .error({
          title: 'Error title',
          body: 'Error body',
          throwCancelError: true
        })
        .result.catch(() => {
          isDismissed = true;
        });

      const wrapper = mount(<ModalDialog />);

      wrapper.find('.rmb-close').simulate('click');
      await flushPromises();

      expect(wrapper.find('.rmb-modal-content').exists()).toBe(false);
      expect(isDismissed).toBe(true);
    });
  });

  describe('type `custom`', () => {
    const openCustomModal = (additionalParams = {}) => {
      return modalService.custom({
        title: 'Custom title',
        body: props => (
          <div className="custom-body">
            <div className="props-for-custom-component">
              {JSON.stringify(
                props,
                (key, value) => {
                  if (typeof value === 'function') {
                    return `function ${key}(){...}`;
                  }

                  return value;
                },
                2
              )}
            </div>
            <button
              type="button"
              className="btn-close"
              // eslint-disable-next-line react/prop-types, react/destructuring-assignment
              onClick={props.closeModal}
            >
              Close Modal
            </button>
          </div>
        ),
        ...additionalParams
      });
    };

    afterEach(async () => {
      await modalService.closeAll();
    });

    it('should match snapshot', () => {
      openCustomModal();

      const tree = renderer.create(<ModalDialog />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should contain modal content', () => {
      modalService.confirm({ title: 'Confirm title', body: 'Confirm body' });

      const wrapper = mount(<ModalDialog />);

      expect(wrapper.find('.rmb-modal-content').exists()).toBe(true);
    });

    it('should be closed', async () => {
      let isClosed = false;

      openCustomModal().result.then(() => {
        isClosed = true;
      });

      const wrapper = mount(<ModalDialog />);

      wrapper.find('.btn-close').simulate('click');
      await flushPromises();

      expect(wrapper.find('.rmb-modal-content').exists()).toBe(false);
      expect(isClosed).toBe(true);
    });

    it('should be dismissed by clicking [X] button', async () => {
      let isDismissed = false;

      openCustomModal({ throwCancelError: true }).result.catch(() => {
        isDismissed = true;
      });

      const wrapper = mount(<ModalDialog />);

      wrapper.find('.rmb-close').simulate('click');
      await flushPromises();

      expect(wrapper.find('.rmb-modal-content').exists()).toBe(false);
      expect(isDismissed).toBe(true);
    });
  });
});
