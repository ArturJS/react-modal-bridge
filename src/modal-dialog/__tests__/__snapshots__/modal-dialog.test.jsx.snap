// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalDialog /> type \`confirm\` should match snapshot 1`] = `
<div
  className="react-modal-mock"
>
  <pre
    className="props"
  >
    {
  "isOpen": true,
  "style": {
    "overlay": {
      "backgroundColor": ""
    }
  },
  "className": "rmb-modal ",
  "shouldCloseOnOverlayClick": true,
  "closeTimeoutMS": 300,
  "contentLabel": "",
  "ariaHideApp": false
}
  </pre>
  <div
    className="TransitionGroup-mock"
  >
    <pre
      className="props"
    >
      {}
    </pre>
    <div
      className="CSSTransition-mock"
    >
      <pre
        className="props"
      >
        {
  "appear": true,
  "timeout": 300,
  "classNames": "rmb-modal-show",
  "mountOnEnter": true,
  "unmountOnExit": true
}
      </pre>
      <div
        className="rmb-modal-content"
      >
        <button
          className="rmb-close"
          onClick={[Function]}
          type="button"
        >
          ×
        </button>
        <div
          className="rmb-modal-header"
        >
          <h3
            className="rmb-modal-title"
          >
            Confirm title
          </h3>
        </div>
        <div>
          <div
            className="rmb-modal-body"
          >
            Confirm body
          </div>
          <div
            className="rmb-modal-footer"
          >
            <button
              className="rmb-btn rmb-btn-primary rmb-btn-ok"
              onClick={[Function]}
              type="button"
            >
              Ok
            </button>
            <button
              className="rmb-btn rmb-btn-default rmb-btn-cancel"
              onClick={[Function]}
              type="button"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ModalDialog /> type \`custom\` should match snapshot 1`] = `
<div
  className="react-modal-mock"
>
  <pre
    className="props"
  >
    {
  "isOpen": true,
  "style": {
    "overlay": {
      "backgroundColor": ""
    }
  },
  "className": "rmb-modal ",
  "shouldCloseOnOverlayClick": true,
  "closeTimeoutMS": 300,
  "contentLabel": "",
  "ariaHideApp": false
}
  </pre>
  <div
    className="TransitionGroup-mock"
  >
    <pre
      className="props"
    >
      {}
    </pre>
    <div
      className="CSSTransition-mock"
    >
      <pre
        className="props"
      >
        {
  "appear": true,
  "timeout": 300,
  "classNames": "rmb-modal-show",
  "mountOnEnter": true,
  "unmountOnExit": true
}
      </pre>
      <div
        className="rmb-modal-content"
      >
        <button
          className="rmb-close"
          onClick={[Function]}
          type="button"
        >
          ×
        </button>
        <div
          className="rmb-modal-header"
        >
          <h3
            className="rmb-modal-title"
          >
            Custom title
          </h3>
        </div>
        <div
          className="rmb-modal-custom-body"
        >
          <div
            className="custom-body"
          >
            <div
              className="props-for-custom-component"
            >
              {
  "closeModal": "function closeModal(){...}"
}
            </div>
            <button
              className="btn-close"
              onClick={[Function]}
              type="button"
            >
              Close Modal
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ModalDialog /> type \`error\` should match snapshot 1`] = `
<div
  className="react-modal-mock"
>
  <pre
    className="props"
  >
    {
  "isOpen": true,
  "style": {
    "overlay": {
      "backgroundColor": "transparent",
      "pointerEvents": "none",
      "zIndex": 1080
    }
  },
  "className": "rmb-modal rmb-modal-error ",
  "shouldCloseOnOverlayClick": true,
  "closeTimeoutMS": 300,
  "contentLabel": "",
  "ariaHideApp": false
}
  </pre>
  <div
    className="TransitionGroup-mock"
  >
    <pre
      className="props"
    >
      {}
    </pre>
    <div
      className="CSSTransition-mock"
    >
      <pre
        className="props"
      >
        {
  "appear": true,
  "timeout": 300,
  "classNames": "rmb-modal-show",
  "mountOnEnter": true,
  "unmountOnExit": true
}
      </pre>
      <div
        className="rmb-modal-content"
      >
        <button
          className="rmb-close"
          onClick={[Function]}
          type="button"
        >
          ×
        </button>
        <div
          className="rmb-modal-header"
        >
          <h3
            className="rmb-modal-title"
          >
            Error title
          </h3>
        </div>
        <div>
          <div
            className="rmb-modal-body"
          >
            Error body
          </div>
          <div
            className="rmb-modal-footer"
          >
            <button
              className="rmb-btn rmb-btn-primary rmb-btn-ok"
              onClick={[Function]}
              type="button"
            >
              Ok
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ModalDialog /> type \`info\` should match snapshot 1`] = `
<div
  className="react-modal-mock"
>
  <pre
    className="props"
  >
    {
  "isOpen": true,
  "style": {
    "overlay": {
      "backgroundColor": "transparent",
      "pointerEvents": "none",
      "zIndex": 1080
    }
  },
  "className": "rmb-modal rmb-modal-info ",
  "shouldCloseOnOverlayClick": true,
  "closeTimeoutMS": 300,
  "contentLabel": "",
  "ariaHideApp": false
}
  </pre>
  <div
    className="TransitionGroup-mock"
  >
    <pre
      className="props"
    >
      {}
    </pre>
    <div
      className="CSSTransition-mock"
    >
      <pre
        className="props"
      >
        {
  "appear": true,
  "timeout": 300,
  "classNames": "rmb-modal-show",
  "mountOnEnter": true,
  "unmountOnExit": true
}
      </pre>
      <div
        className="rmb-modal-content"
      >
        <button
          className="rmb-close"
          onClick={[Function]}
          type="button"
        >
          ×
        </button>
        <div
          className="rmb-modal-header"
        >
          <h3
            className="rmb-modal-title"
          >
            Info title
          </h3>
        </div>
        <div>
          <div
            className="rmb-modal-body"
          >
            Info body
          </div>
          <div
            className="rmb-modal-footer"
          >
            <button
              className="rmb-btn rmb-btn-primary rmb-btn-ok"
              onClick={[Function]}
              type="button"
            >
              Ok
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
